### DISCUSSION TITLE
Is there any lints that fire on refer implicit export from current crate root?

### DISCUSSION THREAD

**Discussion by KisaragiEffective** [rust-lang/rust-clippy#10066] [Q&A]
given those code:
```rs
// src/main.rs
use crate::foo::Foo;
```

```rs
// src/bar.rs
use crate::Foo;
//         ^~~
//         Relying implicit export from crate root
```

is there any lint that fire on refering `crate::Foo`?

**Comment by flip1995** [comment]
Not that I know of. What would be the purpose of this lint? If you would remove the `use` item in `main.rs`, you would get a compiler error in `bar.rs`. And if `Foo` would come from an external crate, then removing a re-export is a breaking change, which shouldn't happen without a major version bump.

The closest thing I can think of is the [`pub_use`](https://rust-lang.github.io/rust-clippy/master/index.html#pub_use) lint, which discourages usage of public re-exports completely.

**Reply by KisaragiEffective** [reply]
> What would be the purpose of this lint?

I think it is mis-leading about their locations, and would like to:
* avoid referencing them 
* refer members by *true* absolute path and not implicit-re-export from root.

> If you would remove the use item in main.rs, you would get a compiler error in bar.rs.

I know this, however my use-case is `bar.rs` is under my control (and can fix manually).

> if `Foo` would come from an external crate, then removing a re-export is a breaking change

I know this too, however I meant `Foo` is *also* in same crate.

**Reply by flip1995** [reply]
Ok, there isn't a lint like this right now. It shouldn't be too hard to implement and would fit in the `restriction` category. But if you'd like to avoid referencing them, why are those re-exports there in the first place? Would a lint disallowing those re-exports make more sense for you? Fight the cause, not the symptoms :thinking:
